### ISSUE TITLE
Adapting for Music AI?

### ISSUE THREAD

**Issue by asigalov61** [drahoslove/pianco#29]
Hey @drahoslove 

First of all, thank you for this project. It is really nice and looks/plays really well.

I was wondering if you would be interested in adapting your project for Music AI? Either as a front-end or stand-alone app?

Here is my Piano Music AI implementation but I desperately need help with front-end / app.

https://github.com/asigalov61/GIGA-Piano

Please let me know if this is something that may interest you.

Thank you.

Sincerely,

Alex

**Comment by drahoslove** [comment]
Hi @asigalov61,
thank you.

I tried your GIGA-piano in the [Colab](https://colab.research.google.com/github/asigalov61/GIGA-Piano/blob/main/GIGA_Piano_Composer.ipynb), and it looks nice.

**I have some questions:**

1) You used the word "desperately". Are you under some pressure? Could you tell me more about the project? 
Is it some university project, part of some business idea, or something you do in your free time, just for your own enjoyment?

2) I see that you also reached out for help to the authors of some other piano projects.
One of them is a web app too, but one is a desktop app, and one is a VTS plugin. So it seems, that you are not really sure what kind of front-end you want. Right?
If you have at least some general idea, could you perhaps elaborate a bit, on what solution you have in mind? What basic features should the UI have?

I'm not very familiar with creating desktop apps. So I won't help with that.
I'm also not very familiar with python, but I can imagine that:
 - you could create a custom web service in python (by wrapping your code into a Django framework or something like that)
 - and implement some well-defined API, which then could be used by some front-end web app implemented by another person.
 
**Regarding me helping you:** 
 
I'd like to keep my Pian.co simple, without adding too many new features and settings.
 
But I can imagine a simplistic integration to my Pian.co I might be interested in. 
It could work like this:
  - there could be another room called `AI`, which would behave similarly to other rooms, but
    - the last few notes played by the user there would be recorded
    - there could be a magical button that would `Let the AI continue on what you just started`
    - clicking on this button would send the recorded notes as a MIDI to your service, it would do the generation and send back the whole piece, which would then be replayed locally. 
    - (it could happen automatically without clicking a button, by just not playing more notes for a while)
 
 
However, if you'd like to have more features, for example, an option to upload MIDI files, a possibility to set various variables,  the ability to save the generated file, etc., then it would be better to have a stand-alone app with UI just for your needs.

**Other thoughts:**

I also have a concern about the performance, if it would be a remote service.
As I tried it in the Colab environment it seemed to take some time for the generating to finish. I'm not sure how well would it run on the server when used by several users at once.

 A downloadable app with your python running locally on the user's device is probably(?) a more feasible solution.
 Can't help you with that, however.
 
So yeah... I'm not sure if I'm interested until you provide me with more info.
 
Feel free to respond to me in an email, if you don't wish to discuss it further publicly.

**Comment by asigalov61** [comment]
@drahoslove Thank you so much for your kind and detailed response,

To answer your questions...

Yes, I reached out to other people with piano projects because you never know who would respond and who would be able to help.

Ideally, I want to have either an app or a front-end. As you are correctly noticed I am not 100% sure yet what can be done because I simply do not have the skills to even guess what would be the best solution for my project.

However, I am willing to work with anyone who would be willing to help me no matter which direction it would go.

This is NOT a business project. It's rather a passion project and I do not have a lot of resources. However, I believe with the right planning and implementation, I should be able to make it happen :)

Now, in regard to your proposal for your PIANCO... I think it is a great idea and I would love to talk to you more about it if you are willing.

Basically, my GIGA-Piano has two models (small and large), and also even smaller models can be made that would run well on most hardware (even w/o GPU). The reason why it was slow in colab was that it was batched generation up-to max seq_len, which is slow and resource intensive. 

So, to expand on your idea for PIANCO... a "magic button" can be made to continue user input notes. If continuation is not long, it will be relatively quick and would only require CPU resources. As I said above, I can reduce the model which will speed it up and so that it can be packaged with your software (app).

I hope this makes sense to you and I hope to chat with you more about it if you are willing. :) Publicly or privatly. I do not mind discussing it here...

Other than that, I really liked your software. It is very pretty and beautiful so I think AI model would really make it shine. :)

Let me know.

Alex

**Comment by asigalov61** [comment]
@drahoslove Ohh...I am really good with python but I am a complete noob with web/app stuff so please forgive my ignorance and misunderstanding in advance :)

However, I would be willing to help however I can and do my part if you are willing to colab and if you think this is something that interests you.

Another interesting idea that was implemented in another project is a real-time accompaniment...user plays melody notes and AI plays chords...

Basically, there is a lot of cool things that can be done :)

**Comment by drahoslove** [comment]
@asigalov61 
Well, that is great.
In that case, let's try the "magic button" idea.

My expectation is, that your python code would not be packaged in my app, but it would exist as a separate microservice, possibly hosted on its own, which would be called by me over http(s).
That way you would be able to utilize the same project in multiple ways. And not be dependent only on my app.
For example in the future, you could have your own landing page, (similar to https://openai.com/blog/musenet/), which could reuse the same python service of yours to present some examples.... or you could let someone create another front-end app, or implement some kind of hardware piano interface from Raspberry Pi/Arduino, or whatever - all of that could use the same service.


**I would like you to try to create some 'proof of concept'.**

Are you familiar with some basics of HTTP and how to create an HTTP handler in python?
Would you be able to create a simple HTTP service with this smaller model you are talking about, that could be called remotely?

It should have an endpoint that would accept the initial notes (and some additional parameters - eg. length of the result) and return the generated continuation.

Let's say the endpoint would be:
 `POST /midi/continuation` 

1) It could either accept and return the data as MIDI file (Standard Midi Format)

    You could mimic the request you'd receive from me using the following curl command (It assumes you are running the service locally at port 8080 and you have file `test_input.mid` containing the initial notes in the current directory):
    ```
    curl -X POST http://localhost:8080/midi/continuation 
       -F notes@test_input.mid
       -F length=15000
    ```

2) Or we can use some custom JSON format (which would be probably a bit easier for me)
   For example:
   ```
   curl -X POST http://localhost:8080/midi/continuation
      -H 'Content-Type: application/json'
      -d '{"notes": [{"note": 60, "time": 0, "duration": 250}, {"note": 64, "time": 1000, "duration": 500}], "length": 15000}'
   ```
   Or less verbose way:
      ```
   curl -X POST http://localhost:8080/midi/continuation
      -H 'Content-Type: application/json'
      -d '{"notes": [[60, 0, 250],[64,1000,500]], "length": 15000}'
   ```
    ... that is for a discussion - I'm not sure what values from the midi file you actually use.

If you have no idea what I am talking about. I could probably help you implement that.
In that case, I'd need some time to familiarize myself with the python language and your code a bit - it's been a while since I've done something in python.

But it would be great if you could handle as much of the python part as possible by yourself, and ideally even host it somewhere by yourself (at least for now, to test it - we could later decide to move it).

**Comment by asigalov61** [comment]
@drahoslove Thank you for your response and thank you for being willing to colab :)

I absolutely agree with you (especially after talking to you and other people) that web API (similar to MuseNet is the way to go. It would be the best and most convenient for everyone and also very versatile in terms of integration.

Thank you for the basic guidance on the API... It is very helpful and I agree very much with most of the examples...I will need to think about which way is the best to go...

Now, I already started working on a prototype for the API... I will host it on paperspace.com as they offer GPU-powered instances. Any help from you would be really appreciated.

I hope you do not mind, I will email you soon with details and examples of my API prototype so that you can take a look at it and help me out if you can...

Otherwise, looking forward to doing this project with you and hopefully, we can achieve some results.

I will close the issue for now and I will email you directly shortly.

Alex

**Comment by drahoslove** [comment]
Awesome!
Looking forward to hearing from you.

**Comment by asigalov61** [comment]
@drahoslove Emailed you @ i@draho.cz with API stuff. Thanks.
